File: 2487.py
class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return None
        head.next = self.removeNodes(head.next)
        if head.next is not None and head.val < head.next.val:
            return head.next
        else:
            return head

File: sol/2487.cpp
struct Solution {
   ListNode *tail(ListNode *node, int &tailmax){
       if (node->next == nullptr){
           tailmax = node->val;
           return node;
       }
       ListNode *tailhead = tail(node->next, tailmax);
       if (node->val >= tailmax){
           node->next = tailhead;
           tailhead = node;
           tailmax = node->val;
       }
       return tailhead;
   }

   inline ListNode* removeNodes(ListNode* head) {
       int tailmax;
       return tail(head, tailmax);
   }
};
